### PR TITLE
Add non-blocking functions for MPMCQueue

### PR DESCRIPTION
**Pull request by gengliqi**
Signed-off-by: gengliqi <gengliqiii@gmail.com>

### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tiflash/issues/5310

Problem Summary:

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

**Review by fuzhe1989 — APPROVED** [review]
great job, LGTM

File: dbms/src/Common/MPMCQueue.h [hunk 1]
```diff
@@ -286,7 +314,10 @@ class MPMCQueue
 
         std::unique_lock lock(mu);
 
-        wait(lock, writer_head, node, pred, deadline);
+        if constexpr (need_wait)
+        {
+            wait(lock, writer_head, node, pred, deadline);
+        }
```
Comment (LittleFall) [inline dbms/src/Common/MPMCQueue.h:321]: ditto with `pred`

**Review by LittleFall — APPROVED** [review]
others LGTM, maybe you can add some inlined overload function to optimize the perfmance

**Pull request merged**